guide.md

```diff
@@ -0,0 +1,324 @@
+Target //src:hello-world up-to-date:
+  bazel-bin/src/hello-world
+INFO: Elapsed time: 0.377s, Critical Path: 0.22s
+INFO: 6 processes: 4 internal, 2 linux-sandbox.
+INFO: Build completed successfully, 6 total actions
+```
+
+# Making Nix optional
+
+It is entirely possible to make the use of Nix optional in a Bazel setup. To do
+so, we use `bazelrc` configs to only enable Nix-specific configuration when a
+`--config` flag is provided, like so:
+
+```
+build:nix --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+run:nix   --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build:nix --crosstool_top=@nixpkgs_config_cc//:toolchain
+```
+
+None of these options will apply to `bazel build` or `bazel run` unless an
+extra `--config nix` parameter is passed:
+
+```
+> bazel run //src:hello-world --config=nix
+...
+```
+
+# Working from a Nix-provided developer environment
+
+Another practical advantage of using Nix is that we can provide consistent,
+zero-setup development environments to all contributors to a project. These
+function similarly to, say, Python virtualenvs, but they're not just limited to
+providing dependencies for just one language ecosystem. Even better, they can
+provide system dependencies too.
+
+```
+> hello
+hello: command not found
+> nix-shell -p hello
+nix-shell> hello
+Hello, world!
+nix-shell> exit
+> hello
+hello: command not found
+```
+
+There's still a few loose ends in our setup that Nix developer environments can
+help tie up. First, the version of Bazel we've been using so far, has been
+unspecified. The general practice is to simply use one's distribution's package
+manager, something that doesn't deliver consistent Bazel versions across major
+OS platforms. Second, we haven't _disabled_ Bazel's deep-rooted tendency to 
+depend on the global environment. Indeed, you can rename `.bazelrc` and rerun
+`bazel run`, and everything will work just fine: Bazel will simply fall back to
+`@local_config_cc`, because a C++ compiler toolchain is still globally available.
+
+We can provide a Nix developer environment that only provides the specific dependencies that developers need:
+
+```nix
+# flake.nix
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.05";
+    flake-compat = {
+      url = github:edolstra/flake-compat;
+      flake = false;
+    };
+    flake-utils.url = github:numtide/flake-utils;
+  };
+  outputs = { nixpkgs, flake-utils, ... }:
+    flake-utils.lib.eachDefaultSystem (system:
+      let pkgs = import nixpkgs { inherit system; };
+      in
+      {
+        devShells.default = with pkgs; mkShellNoCC {
+          name = "rules_nixpkgs_shell";
+          packages = [ bazel_5 bazel-buildtools cacert gcc nix git ];
+        };
+      });
+}
+```
+
+Now developers on different platforms can enter a development environment by
+running `nix develop`. We've also removed the "default" system-wide CC
+toolchain (using `mkShellNoCC` instead of the more common `mkShell`), so
```

Comment:
> We've also removed the "default" system-wide CC
> toolchain (using `mkShellNoCC` instead of the more common `mkShell`),

I think Nix specifics like these will require a bit more expanding.

Reply:
Done.